ibtk/src/utilities/IBTKInit.cpp

```diff
@@ -0,0 +1,79 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2014 - 2019 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "ibtk/IBTKInit.h"
+#include "ibtk/IBTK_MPI.h"
+#include "ibtk/app_namespaces.h"
+
+namespace IBTK
+{
+/////////////////////////////// STATIC //////////////////////////////////////
+#ifdef IBTK_HAVE_LIBMESH
+LibMeshInit* IBTKInit::s_libmesh_init = nullptr;
+#endif
+bool IBTKInit::s_initialized = false;
+
+void
+IBTKInit::initialize(int argc, char** argv, IBTK_MPI::comm communicator, char* petsc_file, char* petsc_help)
+{
+    if (s_initialized)
+    {
+        TBOX_WARNING("IBAMR already initialized. Multiple initialize() calls will be ignored.");
+    }
+    static IBTKInit instance(argc, argv, communicator, petsc_file, petsc_help);
+}
+
+/////////////////////////////// PRIVATE //////////////////////////////////////
+
+IBTKInit::IBTKInit(int argc, char** argv, IBTK_MPI::comm communicator, char* petsc_file, char* petsc_help)
+{
+#ifdef IBTK_HAVE_LIBMESH
+    s_libmesh_init = new LibMeshInit(argc, argv, communicator);
+    libMesh::ReferenceCounter::disable_print_counter_info();
```

Comment:
I'd prefer not to disable this unconditionally - it is occasionally useful to see. Users should disable it on their own if they don't want to see what libMesh has to say.

Reply:
I am not sure, but I think that this may be to prevent emitting some libMesh messages in models that do not explicitly use libMesh.

Reply:
That's a good point - for that reason we should disable it.

Reply:
Can we re-enable it after it is disabled? I like this counter also.

Alternatively, would it be possible to somehow conditionally disable?

Reply:
I suspect that it will be pretty rare for users to actually care so we should recommend they call `libMesh::ReferenceCounter::enable_print_counter_info()` on their own if they want this information.

Reply:
Is there a safe way to lazy-initialize libMesh? Not all IBAMR models will use libMesh, and it seems better not to initialize it if it is not used.

Reply:
I don't think so - I think we are stuck always initializing it.

Reply:
What if we leave application codes to setup `LibMeshInit` or not, and to have them pass in to the constructor if they are using libMesh?

Reply:
This is sort of the status quo now -- we do not automatically setup libMesh (or PETSc for that matter) in application codes.

Is this too complicated: `IBTK_MPI` does not setup PETSc or libMesh, but we provide some convenience classes that can manage standard setup tasks (e.g. `IBTK_PETSc_Initialize` for models that use PETSc but not libMesh, or `IBTK_libMesh_Initialize` for models that use libMesh)?

Reply:
> Is this too complicated: `IBTK_MPI` does not setup PETSc or libMesh, but we provide some convenience classes that can manage standard setup tasks (e.g. `IBTK_PETSc_Initialize` for models that use PETSc but not libMesh, or `IBTK_libMesh_Initialize` for models that use libMesh)?

These would need to be matched with corresponding `finalize` and `shutdown` calls, as well as be called in the correct order. If this is the preferred method, I can switch things around. I don't have strong feelings on it.

Reply:
@drwells what do you think?

I guess that as long as the various libraries "stay out of the way" when they aren't being used, it doesn't really matter if they are activated or not.

Reply:
This sounds too complicated. We should just have one class set up all our dependencies but make it clear in the documentation of that class what exactly it does (i.e., IBAMR does not depend on initialization functions itself, but its dependencies do, so these functions always need to be called).

Reply:
That's what I thought. Thanks! :-)